### PR TITLE
Save weights of CLIP vision encoder and reload them in Multimodal LM

### PR DESCRIPTION
**Pull request by LBerth**
- Add a method to CLIP model to save the vision encoder weights and parameters
- Add a callback to the CLIP lighning module to save the weights at the end of validation epoch if loss improved
- Add option to Multimodal LM to use a resnet vision encoder pretrained with CLIP
- Add unit tests for all features

Note : 
- the vision encoder is saved like this : 
```python
ckpt = {
            "model_state_dict": self.image_encoder.state_dict(),
            "num_channels": self.image_encoder.num_channels,
            "num_classes": self.image_encoder.num_classes,
            "encoder_depth": self.image_encoder.settings.encoder_depth,
            "encoder_stride": self.image_encoder.settings.encoder_stride,
        }
torch.save(ckpt, path)
```
- torch did not like me trying to save directly the ResNet50Settings class in the checkpoint so I saved separately each  argument of  the ResNet50Settings class in the checkpoint
- it was also technically possible to save the whole ResNet50 model but as the pytorch doc says : 
```
The disadvantage of this approach is that the serialized data is bound to the specific classes and the exact directory structure used when the model is saved. The reason for this is because pickle does not save the model class itself. Rather, it saves a path to the file containing the class, which is used during load time. Because of this, your code can break in various ways when used in other projects or after refactors.
```
So I preferred the method of saving the weights and the few arguments we need to instantiate the class.

### CODE REVIEW

File: mfai/torch/lightning_modules/clip.py [hunk 1]
```diff
@@ -116,3 +117,34 @@ def validation_step(
...
+                    # remove the previous checkpoint
+                    dst_folder = Path(trainer.logger.log_dir)
+                    previous_paths = list(dst_folder.glob("visual_encoder_ep-*.tar"))
+                    if previous_paths:
+                        previous_paths[0].unlink()
+
+                    # Store the encoder state dict (weights) and its parameters:
+                    path = dst_folder / f"visual_encoder_ep-{trainer.current_epoch}.tar"
+                    pl_module.model.save_vision_encoder(path)
+                    print("Saved visual encoder weights to", path)
```
Comment (colon3ltocard) [inline mfai/torch/lightning_modules/clip.py:150]: If something wrong happens during saving of the tar the previous tar is laready deleted. Consider this:
1. rename the old file using atomic filesystem operation
2. create the new one
3. delete the old one if and only if the new one was created

File: mfai/config/clip.yaml [hunk 2]
```diff
@@ -29,4 +29,14 @@ trainer:
...
+  callbacks:
+    - class_path: mfai.torch.lightning_modules.clip.SaveCLIPVisualEncoderWeights
+    - class_path: lightning.pytorch.callbacks.ModelCheckpoint
+      init_args:
+        monitor: val_loss
+        filename: "ckpt-{epoch:02d}-{val_loss:.2f}"
+        save_top_k: 1
+  limit_train_batches: 3    # Limit num of batches per epoch (to do very short training)
+  limit_val_batches: 3
+  limit_test_batches: 3
```
[the remark this answers is not in the document]
Comment (LBerth) [reply mfai/config/clip.yaml:42]: Indeed during a full training we would remove them. Its useful during the tests to do a full training on only a few batch, we have also set `max_epochs=2`. I put this here because the `fast_dev_run` option doesn't start the callbacks and logger that I want to test. 
But if you prefer I could put the `limit_X_batches` options in the test command line.

[the remark this answers is not in the document]
Comment (colon3ltocard) [reply mfai/config/clip.yaml:42]: I understand thanks. Yep I think the yaml should be ready to train/use I like your proposal of moving the limit_X to the test code.

**Review by colon3ltocard — DISMISSED** [review]

File: mfai/torch/lightning_modules/clip.py [hunk 3]
```diff
@@ -116,3 +117,43 @@ def validation_step(
...
+                    epoch = trainer.current_epoch
+                    ckpt_path = dst_folder / f"visual_encoder_ep-{epoch}.tar"
+                    pl_module.model.save_vision_encoder(ckpt_path)
+                    print("Saved visual encoder weights to", ckpt_path)
+
+                    # remove the old checkpoints
+                    if ckpt_path.exists():
+                        old_ckpts = list(dst_folder.glob("visual_encoder_ep-*_old.tar"))
+                        for old_ckpt in old_ckpts:
+                            old_ckpt.unlink()
```
Comment (kalharko) [inline mfai/torch/lightning_modules/clip.py:159]: This code block can be simplified by changing execution conditions to exit conditions:
```python
if current_val_loss >= self.best_val_loss:
    return

if trainer.logger is None:
    return
    
if not isinstance(pl_module, CLIPLightningnModule:
    return

if not trainer.logger.log_dir:
    return  # Maybe include a warning here ?
```

File: mfai/torch/models/resnet.py [hunk 4]
```diff
@@ -19,7 +19,10 @@ class ResNetEncoder(ResNet):
     - output channels specification of feature tensors (produced by encoder)
     - patching first convolution for arbitrary input channels
     """
-    def __init__(self, out_channels: tuple[int, ...], depth: int = 5, **kwargs: dict[str, Any]):
+
+    def __init__(
+        self, out_channels: tuple[int, ...], depth: int = 5, **kwargs: dict[str, Any]
```
Comment (kalharko) [inline mfai/torch/models/resnet.py:24]: mypy is smart enought to know that `**kwargs` means `dict[str, <something>]`. Type hinting `Any` is sufficient.
```suggestion
        self, out_channels: tuple[int, ...], depth: int = 5, **kwargs: Any
```

File: mfai/torch/models/resnet.py [hunk 5]
```diff
@@ -45,11 +48,13 @@ def forward(self, x: torch.Tensor) -> list[torch.Tensor]:
 
         return features
 
-    def load_state_dict(self, state_dict: dict[str, Any], **kwargs: dict[str, Any]) -> None:
+    def load_state_dict(
+        self, state_dict: dict[str, Any], **kwargs: dict[str, Any]
```
Comment (kalharko) [inline mfai/torch/models/resnet.py:52]:
```suggestion
        self, state_dict: dict[str, Any], **kwargs: Any
```

**Review by tourniert — APPROVED** [review]
